File: javascript/test/ssrf.test.js
const { isPrivateOrReservedIp, isPublicHttpUrl } = require('../src/ssrf');

test('isPrivateOrReservedIp flags private/reserved addresses', () => {
  for (const ip of ['127.0.0.1', '10.0.0.1', '192.168.1.1', '169.254.169.254', '::1', 'fe80::1']) {
    expect(isPrivateOrReservedIp(ip)).toBe(true);
  }
});

test('isPrivateOrReservedIp allows public addresses', () => {
  for (const ip of ['8.8.8.8', '1.1.1.1']) {
    expect(isPrivateOrReservedIp(ip)).toBe(false);
  }
});

test('isPublicHttpUrl rejects non-http(s) scheme without DNS lookup', async () => {
  expect(await isPublicHttpUrl('javascript:alert(1)')).toBe(false);
});

test('isPublicHttpUrl rejects localhost', async () => {
  expect(await isPublicHttpUrl('http://localhost/')).toBe(false);
  expect(await isPublicHttpUrl('http://127.0.0.1/')).toBe(false);
});

File: go/ssrf_test.go
package securekit

import (
	"net"
	"testing"
)

func TestIsPrivateOrReservedIP(t *testing.T) {
	private := []string{"127.0.0.1", "10.0.0.1", "192.168.1.1", "169.254.169.254", "::1", "fe80::1"}
	public := []string{"8.8.8.8", "1.1.1.1"}

	for _, s := range private {
		if !IsPrivateOrReservedIP(net.ParseIP(s)) {
			t.Errorf("expected %q to be flagged private/reserved", s)
		}
	}
	for _, s := range public {
		if IsPrivateOrReservedIP(net.ParseIP(s)) {
			t.Errorf("expected %q to be flagged public", s)
		}
	}
}

func TestIsPublicHTTPURLRejectsNonHTTPScheme(t *testing.T) {
	if IsPublicHTTPURL("javascript:alert(1)") {
		t.Fatal("expected non-http(s) scheme to be rejected without any DNS lookup")
	}
}

func TestIsPublicHTTPURLRejectsLocalhost(t *testing.T) {
	if IsPublicHTTPURL("http://localhost/") {
		t.Fatal("expected localhost to resolve to loopback and be rejected")
	}
	if IsPublicHTTPURL("http://127.0.0.1/") {
		t.Fatal("expected raw loopback IP to be rejected")
	}
}
